gh-98731: Improvements to the logging documentation (#101618)
python/cpython: Doc/howto/logging.rst

^^^^^^^^^^^^^^^^^^^

> Logging provides a set of convenience functions for simple logging usage. These

Reviewer: Since your change refers to using logger methods rather than the convenience functions, I'd remove this sentence and go with something like "You can access logging functionality by creating a logger via ..."

Author: Done

can be accessed by creating a logger via ``logger = getLogger(__name__)``, and

> then calling :func:`logger.debug`, :func:`logger.info`, :func:`logger.warning`,

Reviewer: Use `: meth:` here rather than `:func:` maybe? This applies to similar usages below.

Author: Done

:func:`logger.error` and :func:`logger.critical`. To determine when to use

> logging, see the table below, which states, for each of a set of common tasks,

Reviewer: ```suggestion
logging, and to see which logger methods to use when, see the table below. It states, for each of a set of common tasks,
```

Author: Done

> the best tool to use for it.

Reviewer: ```suggestion
the best tool to use for that task.
```

Author: Done

> +-------------------------------------+--------------------------------------+

Reviewer: Did you actually build this documentation locally? If you did, not sure how it got past these errors.

Author: I did not. I will try to build them locally and message if I have issues.

|                                     |                                      |
|                                     | :func:`logger.warning` if there is  |
|                                     | nothing the client application can do|
|                                     | about the situation, but the event   |
|                                     | should still be noted                |
+-------------------------------------+--------------------------------------+
| Report an error regarding a         | Raise an exception                   |
| particular runtime event            |                                      |
+-------------------------------------+--------------------------------------+
| Report suppression of an error      | :func:`logger.error`,               |
| without raising an exception (e.g.  | :func:`logger.exception` or         |
| error handler in a long-running     | :func:`logger.critical` as          |
| server process)                     | appropriate for the specific error   |
|                                     | and application domain               |
+-------------------------------------+--------------------------------------+

> The functions are named after the level or severity of the events

Reviewer: ```suggestion
The logger methods are named after the level or severity of the events
```

Author: Done

   WARNING:root:Watch out!

printed out on the console. The ``INFO`` message doesn't appear because the
default level is ``WARNING``. The printed message includes the indication of the
level and the description of the event provided in the logging call, i.e.
'Watch out!'. The actual output can be formatted quite flexibly if you need
that; formatting options will also be explained later.

Notice that in this example, we use functions directly on the ``logging``
module, like ``logging.debug``, rather than creating a logger and calling
functions on it. These functions operation on the root logger, but can be useful

> as they will call ``basicConfig`` for you if it has not been called yet, like in

Reviewer: ```suggestion
as they will call :func:`~logging.basicConfig` for you if it has not been called yet, like in
```

Author: Done

> this example.  In larger programs you'll usually want to control this call

Reviewer: ```suggestion
this example.  In larger programs you'll usually want to control the logging configuration
```

Author: Done

> explicitly however, so for that reason as well as others, it's better to explicitly

Reviewer: ```suggestion
explicitly, however - so for that reason as well as others, it's better to
```

Author: Done

> create loggers and call their member functions.

Reviewer: ```suggestion
create loggers and call their methods.
```

Author: Done

   logging.basicConfig(level=numeric_level, ...)

The call to :func:`basicConfig` should come *before* any calls to

> :func:`logger.debug`, :func:`logger.info`, etc. Otherwise, that logging

Reviewer: ```suggestion
:func:`logger.debug`, :func:`logger.info`, etc. Otherwise, that logging event
```

Author: Done

> will not be handled.

Reviewer: Not strictly true. For `WARNING` and higher, a handler of last resort outputs the event to `sys.stderr`.

Author: I only discovered this behavior recently 😛 (and have mixed feelings on it). I adjusted it to be technically correct without getting into the details of the last resort handler; but if you would prefer I mention it I can do that as well.
